src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/FileServerHandler.java

```diff
@@ -265,23 +268,151 @@ private void serveDefaultFavIcon(HttpExchange exchange, boolean writeBody)
+            long start, end;
+            try {
+                if (startStr.isEmpty()) {  // "-<suffix-length>"
+                    long suffixLength = Long.parseLong(endStr);
+                    if (suffixLength <= 0) {
+                        return null;
+                    }
+                    start = Math.max(0, fileSize - suffixLength);
+                    end = fileSize - 1;
+                } else {  // "<start>-<end>" or "<start>-"
+                    start = Long.parseLong(startStr);
+                    end = endStr.isEmpty() ? fileSize - 1 : Long.parseLong(endStr);
+                    if (end >= fileSize) {
+                        end = fileSize - 1;
+                    }
+                }
+            } catch (NumberFormatException e) {
+                return null;  // invalid number format
+            }
+
+            if (start < 0 || start > end)
+                return null;  // invalid range values
+
+            ranges.add(new RangeEntry(start, end));
+        }
+        return ranges;
+    }
+
+    private void servePartialContents(HttpExchange exchange, Path path, List<RangeEntry> ranges)
+            throws IOException
+    {
+        var respHdrs = exchange.getResponseHeaders();
+        String fileContentType = mediaType(path.toString());
+        String boundary = UUID.randomUUID().toString();
+        long fileSize = Files.size(path);
+        boolean isSingleRange = ranges.size() == 1;
+        long responseLength;
+        if (isSingleRange) {
+            RangeEntry range = ranges.get(0);
+            respHdrs.set("Content-Range",
+                    "bytes %s-%s/%s".formatted(range.start, range.end, fileSize));
+            responseLength = range.end - range.start + 1;
+        } else {
+            respHdrs.set("Content-Type", "multipart/byteranges; boundary=" + boundary);
+            responseLength = 0;  // 0 for chunked transfer encoding
+        }
+        exchange.sendResponseHeaders(206, responseLength); // Partial Content
+        try (RandomAccessFile raf = new RandomAccessFile(path.toFile(), "r");
+             OutputStream os = exchange.getResponseBody()) {
+            for (RangeEntry range : ranges) {
+                if (!isSingleRange) {
+                    os.write(("--" + boundary + "\r\n").getBytes(UTF_8));
+                    os.write(("Content-Type: " + fileContentType + "\r\n").getBytes(UTF_8));
+                    os.write("Content-Range: bytes %s-%s/%s\r\n\r\n"
+                            .formatted(range.start, range.end, fileSize).getBytes(UTF_8));
+                }
+                raf.seek(range.start);
+                long bytesToWrite = range.end - range.start + 1;
+                byte[] buffer = new byte[8192];
+                while (bytesToWrite > 0) {
+                    int len = raf.read(buffer, 0, (int) Math.min(buffer.length, bytesToWrite));
+                    if (len == -1) {
+                        break;  // EOF
```

Comment:
Is this not an error situation, if we reach EOF and have only partially returned the requested range?

Reply:
I'll have to check the RFC but it might be allowable to return a shorter range than was requested, but the actual range to be returned would have to have been specified in the `Content-Range` header. You would have to check the length of the file first. So, EOF wouldn't occur except in very unusual circumstances. Otherwise, yes throwing IOE would close the connection signaling the error.